wazuh/wazuh
# Compare the architecture when evaluating vulnerable packages for RHEL

## Description

This PR closes #6212. Now, when evaluating Redhat vulnerabilities the architecture of packages is taken into account when matching vulnerabilities, following the next criteria:

- Packages with `noarch` architecture match any. Based on the `noarch` definition:

_Used by the RPM Package Manager to mark software packages that contain only content, such as graphics, documentation, or similar data that can be used on any architecture._

- When affected packages have no fixed version, the architecture of the vulnerable package is not included in the feed. In these cases, the architecture is not checked.

- Otherwise, if the architecture of the installed package and the vulnerable one doesn't match, the vulnerability is discarded.

The same behavior cannot be applied to the rest of the feeds since:

- The NVD doesn't report the architecture of vulnerable packages in most cases:

```
sqlite> select count(*),TARGET_HW from NVD_CPE group by TARGET_HW;
209779|*
3|arm
30|arm64
1|x32
192|x64
184|x86
```

- Canonical and Debian don't include affected architectures in their OVAL feeds.

## Manual tests

To ensure the correct operation of the fix, we are studying some use cases with the same CVE and package that the detailed in the issue.

### Test 1:

- **kernel version:** 3.10.0-229.7.2.el7
- **kernel arch:** x86_64

**Result**

The package is vulnerable since the fix for `x86_64` was introduced in `3.10.0-327.36.3.el7`.

```
2020/10/08 15:44:21 wazuh-modulesd:vulnerability-detector[12513] wm_vuln_detector.c:1946 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5458): Package 'kernel' inserted into the vulnerability 'CVE-2016-5195'. Version (3.10.0-229.7.2.el7) 'less than' '3.10.0-327.36.3.el7' (feed 'OVAL').
```

### Test 2:

- **kernel version:** 3.10.0-328.36.3.el7
- **kernel arch:** x86_64

**Result**

Now, the package is not vulnerable since its version is higher than `3.10.0-327.36.3.el7`. It doesn't match with the `ARM` vulnerable package as expected.

```
2020/10/08 16:33:24 wazuh-modulesd:vulnerability-detector[15680] wm_vuln_detector.c:1902 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5460): Package 'kernel' not vulnerable to 'CVE-2016-5195'. Version (3.10.0-328.36.3.el7) not 'less than' '3.10.0-327.36.3.el7' (feed 'OVAL').
```

### Test 3:

- **kernel version:** 3.10.0-328.36.3.el7
- **kernel arch:** aarch64

**Result**

Now, we change the package architecture to `aarch64`. With the same version as _Test 2_, now the package is marked as vulnerable.

```
2020/10/08 16:42:05 wazuh-modulesd:vulnerability-detector[16292] wm_vuln_detector.c:1946 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5458): Package 'kernel' inserted into the vulnerability 'CVE-2016-5195'. Version (3.10.0-328.36.3.el7) 'less than' '4.5.0-15.2.1.el7' (feed 'OVAL').
```

### Test 4:

- **kernel version:** 3.10.0-229.7.2.el7
- **kernel arch:** noarch

**Result**

Now, the version is less than the vulnerable for both architectures, and the architecture is `noarch`. Therefore, it should match with both vulnerable packages.

```
2020/10/09 03:44:05 wazuh-modulesd:vulnerability-detector[46902] wm_vuln_detector.c:1946 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5458): Package 'kernel' inserted into the vulnerability 'CVE-2016-5195'. Version (3.10.0-229.7.2.el7) 'less than' '4.5.0-15.2.1.el7' (feed 'OVAL').
2020/10/09 03:44:05 wazuh-modulesd:vulnerability-detector[46902] wm_vuln_detector.c:1941 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5459): Trying to insert duplicated package 'kernel' into the vulnerability 'CVE-2016-5195'. Version (3.10.0-229.7.2.el7) 'less than' '3.10.0-327.36.3.el7' (feed 'OVAL').
```

## Default tests




- Compilation without warnings in every supported platform


- Memory tests for Linux


- [x] Retrocompatibility with older Wazuh versions
- [x] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)
- [x] Stress test for affected components
